doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,326 @@
+ * read-after-write safety unless read replicas used
+ * Valid keys are `\A[-/_a-zA-Z0-9]+\z` after encoding
+ * Valid buckets are `^[a-zA-Z0-9_-]+$`
+ * Custom Stream Names and Stream ingest subjects to cater for different domains, mirrors and imports
+ * Key starting with `_kv` is reserved for internal use
+ * CLI tool to manage the system as part of `nats`, compatible with client implementations
+ * Read-Only and Read-Write interfaces
+
+Planned features:
+
+ * Ability to read from regional read replicas maintained by JS Mirroring (planned)
+ * A read-only client can talk directly, and only, to a regional read replica
+ * Discoverability of read replicas per cluster based on in-bucket configuration values
+ * Standard encoders/decoders to deliver zero-trust end to end encryption, perhaps using nkeys
+ * Pluggable storage backends
+
+## Data Types
+
+### Result
+
+This is the result made available over watchers, `Get()` etc. All backends must return an implementation providing at least this information.
+
+```go
+type Result interface {
+	// Bucket is the bucket the data was loaded from
+	Bucket() string
+	// Key is the key that was retrieved
+	Key() string
+	// Value is the retrieved value
+	Value() string
+	// Created is the time the data was received in the bucket
+	Created() time.Time
+	// Sequence is a unique sequence for this value
+	Sequence() uint64
+	// Delta is distance from the latest value. If history is enabled this is effectively the index of the historical value, 0 for latest, 1 for most recent etc.
+	Delta() uint64
+	// OriginCluster is the cluster where this data originate from, may be empty if sharing was not enabled
+	OriginCluster() string
+	// Operation is the kind of operation this result represents, enum of PUT and DEL
+	Operation() Operation
+}
+```
+
+### Status
+
+This is the status of the KV as a whole
+
+```go
+type Status interface {
+	// Bucket the name of the bucket
+	Bucket() string
+
+	// Values is how many messages are in the bucket, including historical values
+	Values() uint64
+
+	// History returns the configured history kept per key
+	History() int64
+
+	// TTL is how long the bucket keeps values for
+	TTL() time.Duration
+
+	// Cluster returns the name of the cluster holding the read replica of the data
+	Cluster() string
+
+	// Replicas returns how many times data in the bucket is replicated at storage
+	Replicas() (ok int, failed int)
+
+	// Keys returns a list of all keys in the bucket - not possible now
+	Keys() ([]string, error)
+
+	// BackingStore is a backend specific name for the underlying storage - eg. stream name
+	BackingStore() string
+
+	// MirrorStatus is the status of a read replica, error when not accessing a replica (planned)
+	MirrorStatus() (lag int64, active bool, err error)
+}
+```
+
+## RoKV
+
+This is a read-only KV store handle, every backend should implement a language equivalent interface
+
+```go
+// RoKV is a read-only interface to a single key-value store bucket
```

Comment:
I disagree, it has to be specific because I dont want on a RO KV the `Put()` to be possible at all, thats what interfaces are for.

Reply:
That feels like a permission thing, not a backend impl thing. Feels like it might be over complicating things.

Reply:
Interfaces in go are permissions.

Reply:
They can be, but in our world subjects are and we should always rely on those for the last say so to speak, at least imo.

Reply:
As a programmer, when I configure the client package to be read only, returning a subset interface with only reading methods on it communicates the intent - and abilities even during my time in the IDE.  It informats auto completion and more.

Why do we have io.Reader and io.ReadCloser? dont we just need io.ReadCloser when you can just ignore the Close() method? Because the language design tells you, your IDE and the compiler, that you are never going to or have to close this thing.  Likewise returning a read only interface means given 500 compatible backends - none of them will inappropriately expose write methods.

But again - language client details, ignore this if you like

Reply:
I clarified in the push why I called out the Ro and normal versions of the interface.